site/en/blog/wasmgc/index.md

```diff
@@ -0,0 +1,128 @@
+{% Aside %}
+Reference counting is used in PHP, but most modern browsers now don't use reference-counting for garbage collection.
+{% endAside %}
+
+There are other challenges with garbage collection, like [detecting cycles](https://www.php.net/manual/en/features.gc.collecting-cycles.php), but for this article, having a basic level of understanding of reference counting is enough.
+
+## Programming languages are implemented in other programming languages
+
+It may feel like inception, but programming languages are implemented in other programming languages. For example, the PHP runtime is primarily implemented in C. You can check out the [PHP source code on GitHub](https://github.com/php/php-src/). PHP's garbage collection code is mainly located in the file [`zend_gc.c`](https://github.com/php/php-src/blob/master/Zend/zend_gc.c). Most developers will install PHP via the package manager of their operating system. But developers can also [build PHP from the source code](https://github.com/php/php-src/tree/master#building-php-source-code). For example, in a Linux environment, the steps `./buildconf && ./configure && make` would build PHP for the Linux runtime. But this also means that the PHP runtime can be compiled for other runtimes, like, you guessed it, Wasm.
+
+## Traditional methods of porting languages to the Wasm runtime
+
+Independently from the platform PHP is running on, PHP scripts are compiled into the same bytecode and run by the [Zend Engine](https://en.wikipedia.org/wiki/Zend_Engine). The Zend Engine is a compiler and runtime environment for the PHP scripting language. It consists of the Zend Virtual Machine (VM), which is composed of the Zend Compiler and the Zend Executor. Languages like PHP that are implemented in other high-level languages like C commonly have optimizations that target specific architectures, such as Intel or ARM, and require a different backend for each architecture. In this context, Wasm represents a new architecture. If the VM has architecture-specific code, like just-in-time (JIT) or ahead-of-time (AOT) compilation, then the developer also implements a backend for JIT/AOT for the new architecture. This approach makes a lot of sense because often the main part of the codebase can just be recompiled for each new architecture.
+Given how low-level Wasm is, it is natural to try the same approach there: Recompile the main VM code with its parser, library support, garbage collection, and optimizer to Wasm, and implement a JIT or AOT backend for Wasm if needed. This has been possible since the Wasm MVP, and it works well in practice in many cases. In fact, [PHP compiled to Wasm](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/php-wasm/compile/build.js) is what powers the [WordPress Playground](https://playground.wordpress.net/). Learn more about the project in the article [Build in-browser WordPress experiences with WordPress Playground and WebAssembly](https://web.dev/wordpress-playground/).
```

Comment:
```suggestion

Given how low-level Wasm is, it is natural to try the same approach there: Recompile the main VM code with its parser, library support, garbage collection, and optimizer to Wasm, and implement a JIT or AOT backend for Wasm if needed. This has been possible since the Wasm MVP, and it works well in practice in many cases. In fact, [PHP compiled to Wasm](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/php-wasm/compile/build.js) is what powers the [WordPress Playground](https://playground.wordpress.net/). Learn more about the project in the article [Build in-browser WordPress experiences with WordPress Playground and WebAssembly](https://web.dev/wordpress-playground/).
```